src/charm.py

```diff
@@ -0,0 +1,192 @@
+
+        certificates = self.on["certificates"]
+        self.framework.observe(certificates.relation_broken, self._check_config)
+        self.framework.observe(certificates.relation_created, self._check_config)
+        self.framework.observe(certificates.relation_changed, self._check_config)
+
+        self.framework.observe(self.on.start, self._on_start)
+        self.framework.observe(self.on.test_action, self._on_test_action)
+        self.framework.observe(self.on.config_changed, self._check_config)
+        self.framework.observe(self.on.upgrade_charm, self._upgrade_charm)
+
+    def _kube_control(self, event: EventBase):
+        self.kube_control.set_auth_request(self.unit.name)
+        return self._check_config(event)
+
+    def _check_kube_control(self, event: EventBase) -> bool:
+        self.unit.status = MaintenanceStatus("Evaluating kubernetes authentication.")
+        evaluation = self.kube_control.evaluate_relation(event)
+        if evaluation:
+            if "Waiting" in evaluation:
+                self.unit.status = WaitingStatus(evaluation)
+            else:
+                self.unit.status = BlockedStatus(evaluation)
+            return False
+        if not self.kube_control.get_auth_credentials(self.unit.name):
+            self.unit.status = WaitingStatus("Waiting for kube-control: unit credentials.")
+            return False
+        self.unit.status = MaintenanceStatus("Kubernetes authentication completed.")
+        self.kube_control.create_kubeconfig(
+            self.CA_CERT_PATH, "/root/.kube/config", "root", self.unit.name
+        )
+        self.kube_control.create_kubeconfig(
+            self.CA_CERT_PATH, "/home/ubuntu/.kube/config", "ubuntu", self.unit.name
+        )
+        return True
+
+    def _check_certificates(self, event: EventBase) -> bool:
+        self.unit.status = MaintenanceStatus("Evaluating certificates.")
+        evaluation = self.certificates.evaluate_relation(event)
+        if evaluation:
+            if "Waiting" in evaluation:
+                self.unit.status = WaitingStatus(evaluation)
+            else:
+                self.unit.status = BlockedStatus(evaluation)
+            return False
+        self.CA_CERT_PATH.write_text(self.certificates.ca)
+        return True
+
+    def _check_config(self, event: EventBase) -> None:  # TODO: Not the best name
+        if not self._check_certificates(event):
+            return
+
+        if not self._check_kube_control(event):
+            return
+
+        channel = self.config.get("channel")
+        self._install_snaps(channel)
+
+        self.unit.status = ops.ActiveStatus("Ready to test.")
+
+    def _upgrade_charm(self, event: EventBase) -> None:
+        channel = self.config.get("channel")
+        self._install_snaps(channel)
+
+    def _install_snaps(self, channel: Optional[str]) -> None:
+        self.unit.status = ops.MaintenanceStatus("Installing core snap.")
+        self._ensure_snap("core")
```

Comment:
Did you check if this snap is required?

Reply:
It's not, integration tests and e2e tests pass with this snap removed. I removed it.